### PR TITLE
Improve remind an PW request / reset messages

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/15307

### Summary of Changes

- Make sure when we are not in debug mode there is a "general" message displayed.
- On PW reset atempts for SU add the docs page & send the admin a mail explaining why he cant do this and how to do it.

This has been reported to the JSST by @brianteeman but has been decided to do it in the public tracker as this is a longstanding and public issue anyway. cc @joomla/security 

### Testing Instructions

- install a site whith this patch applyed
- create an non-superuser
- try to reset the non-superuser
- all should work correctly but with a different message
- try to reset the superuser
- there is now the same message but in the background a new mail is send to the super user telling him that he cant do what he wanted to do but how to do it.
- enable joomla debug mode (and also make sure error reporting is active.
- try to reset the super admin again
- you now get the `old` error message but with the docs linked.

### Actual result BEFORE applying this Pull Request

Different messages for different cases that could be used to do `user emuration`

### Expected result AFTER applying this Pull Request

One general message used for all cases
In the special case of the SU there is a mail in the background to inform him about the steps he has to take.

### Documentation Changes Required

None.

### CODE REVIEW

**Comment by zero-24** [comment]
hmm drone fails unrealted on an message that the dev dependencies can not be installed because the minimum PHP Version is 5.3.10.. Any idea how to overcome that error? @HLeithner @Hackwar ? https://ci.joomla.org/joomla/joomla-cms/35781/1/2

File: language/en-GB/en-GB.com_users.ini [hunk 1]
```diff
@@ -138,10 +140,12 @@ COM_USERS_REMIND_DEFAULT_LABEL="Please enter the email address associated with y
 COM_USERS_REMIND_EMAIL_LABEL="Your Email"
 COM_USERS_REMIND_LIMIT_ERROR_N_HOURS="You have exceeded the maximum number of password resets allowed. Please try again in %s hours."
 COM_USERS_REMIND_LIMIT_ERROR_N_HOURS_1="You have exceeded the maximum number of password resets allowed. Please try again in one hour."
+COM_USERS_REMIND_REQUEST="If the email address you entered is registered on this site you will shortly receive an email with a reminder has been send."
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_users.ini]:
```suggestion
COM_USERS_REMIND_REQUEST="If the email address you entered is registered on this site you will shortly receive an email with a reminder."
```

File: language/en-GB/en-GB.com_users.ini [hunk 2]
```diff
@@ -138,10 +140,12 @@ COM_USERS_REMIND_DEFAULT_LABEL="Please enter the email address associated with y
...
 COM_USERS_REMIND_REQUEST_ERROR="Error requesting password reminder."
 COM_USERS_REMIND_REQUEST_FAILED="Reminder failed: %s"
 COM_USERS_REMIND_REQUEST_SUCCESS="Reminder sent. Please check your mail."
-COM_USERS_REMIND_SUPERADMIN_ERROR="A Super User can't request a password reminder. Please contact another Super User or use an alternative method."
+COM_USERS_REMIND_SUPERADMIN_BODY="Hello,\n\na request has been made to reset your %s account password. But, for security reasons, a Super User can't request a password reminder. Please contact another Super User or use an alternative method as described in the documentation: https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F"
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_users.ini]:
```suggestion
COM_USERS_REMIND_SUPERADMIN_BODY="Hello,\n\nA request has been made to reset your %s account password. But, for security reasons, a Super User can't request a password reminder. Please contact another Super User or use an alternative method as described in the documentation: https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F"
```

File: language/en-GB/en-GB.com_users.ini [hunk 3]
```diff
@@ -151,6 +155,7 @@ COM_USERS_RESET_CONFIRM_ERROR="Error while confirming the password."
 COM_USERS_RESET_CONFIRM_FAILED="Your password reset confirmation failed because the verification code was invalid. %s"
 COM_USERS_RESET_CONFIRM_LABEL="An email has been sent to your email address. The email has a verification code, please paste the verification code in the field below to prove that you are the owner of this account."
 COM_USERS_RESET_COMPLETE_TOKENS_MISSING="Your password reset confirmation failed because the verification code was missing."
+COM_USERS_RESET_REQUEST="If the email address you entered is registered on this site you will shortly receive an email with a link to create a password for your account."
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_users.ini]:
```suggestion
COM_USERS_RESET_REQUEST="If the email address you entered is registered on this site you will shortly receive an email with a link to reset the password for your account."
```

File: components/com_users/controllers/remind.php [hunk 4]
```diff
@@ -46,14 +46,11 @@ public function remind()
...
-			// The request succeeded.
-			// Proceed to step two.
-			$message = JText::_('COM_USERS_REMIND_REQUEST_SUCCESS');
-			$this->setRedirect(JRoute::_('index.php?option=com_users&view=login', false), $message);
 
-			return true;
-		}
+		// To not expose if the user exists or not we send a generic message.
+		$message = JText::_('COM_USERS_REMIND_REQUEST');
+		$this->setRedirect(JRoute::_('index.php?option=com_users&view=login', false), $message);
```
Comment (brianteeman) [inline components/com_users/controllers/remind.php]:
```suggestion
		$this->setRedirect(JRoute::_('index.php?option=com_users&view=login', false), $message, 'notice');
```

Display as a blue notice not a green message for consistency with reset request

**Comment by zero-24** [comment]
I have just applied the suggestions thanks @brianteeman

File: language/en-GB/en-GB.com_users.ini [hunk 5]
```diff
@@ -138,10 +140,12 @@ COM_USERS_REMIND_DEFAULT_LABEL="Please enter the email address associated with y
...
+COM_USERS_REMIND_REQUEST="If the email address you entered is registered on this site you will shortly receive an email with a reminder."
 COM_USERS_REMIND_REQUEST_ERROR="Error requesting password reminder."
 COM_USERS_REMIND_REQUEST_FAILED="Reminder failed: %s"
 COM_USERS_REMIND_REQUEST_SUCCESS="Reminder sent. Please check your mail."
-COM_USERS_REMIND_SUPERADMIN_ERROR="A Super User can't request a password reminder. Please contact another Super User or use an alternative method."
+COM_USERS_REMIND_SUPERADMIN_BODY="Hello,\n\nA request has been made to reset your %s account password. But, for security reasons, a Super User can't request a password reminder. Please contact another Super User or use an alternative method as described in the documentation: https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F"
+COM_USERS_REMIND_SUPERADMIN_ERROR="A Super User can't request a password reminder. Please contact another Super User or use an <a href='https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F' target='_blank' rel='noopener noreferrer'>alternative method as described in the documentation</a>."
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_users.ini]:
```suggestion
COM_USERS_REMIND_SUPERADMIN_ERROR="A Super User can't request a password reminder. Please contact another Super User or use an <a href='https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F' target='_blank' rel='noopener noreferrer'>alternative method</a> as described in the documentation."
```

File: language/en-GB/en-GB.com_users.ini [hunk 6]
```diff
@@ -138,10 +140,12 @@ COM_USERS_REMIND_DEFAULT_LABEL="Please enter the email address associated with y
 COM_USERS_REMIND_EMAIL_LABEL="Your Email"
 COM_USERS_REMIND_LIMIT_ERROR_N_HOURS="You have exceeded the maximum number of password resets allowed. Please try again in %s hours."
 COM_USERS_REMIND_LIMIT_ERROR_N_HOURS_1="You have exceeded the maximum number of password resets allowed. Please try again in one hour."
+COM_USERS_REMIND_REQUEST="If the email address you entered is registered on this site you will shortly receive an email with a reminder."
 COM_USERS_REMIND_REQUEST_ERROR="Error requesting password reminder."
 COM_USERS_REMIND_REQUEST_FAILED="Reminder failed: %s"
 COM_USERS_REMIND_REQUEST_SUCCESS="Reminder sent. Please check your mail."
-COM_USERS_REMIND_SUPERADMIN_ERROR="A Super User can't request a password reminder. Please contact another Super User or use an alternative method."
+COM_USERS_REMIND_SUPERADMIN_BODY="Hello,\n\nA request has been made to reset your %s account password. But, for security reasons, a Super User can't request a password reminder. Please contact another Super User or use an alternative method as described in the documentation: https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F"
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_users.ini]: was it intentional that this is not a link?

Comment (zero-24) [reply language/en-GB/en-GB.com_users.ini]: yes as this is a text only mail.

File: language/en-GB/en-GB.com_users.ini [hunk 7]
```diff
@@ -30,6 +30,8 @@ COM_USERS_ERROR_SECRET_CODE_WITHOUT_TFA="You have entered a Secret Code but two
 COM_USERS_FIELD_PASSWORD_RESET_DESC="Please enter the email address associated with your User account.<br />A verification code will be sent to you. Once you have received the verification code, you will be able to choose a new password for your account."
 COM_USERS_FIELD_PASSWORD_RESET_LABEL="Email Address"
 COM_USERS_FIELD_REMIND_EMAIL_DESC="Please enter the email address associated with your User account.<br />Your username will be emailed to the email address on file."
+COM_USERS_REMIND_REQUEST="If the email address was correct a reminder has been sent. Please check your mail."
+COM_USERS_RESET_REQUEST="If the email address was correct a new password has been sent. Please check your mail."
```
Comment (richard67) [inline language/en-GB/en-GB.com_users.ini]: Adding these language strings make them duplicate, i.e. there are 2 with the same key of each in the same language file.

Comment (richard67) [reply language/en-GB/en-GB.com_users.ini]: It seems they once should have been moved down for alphabetical order but have been copied instead, i.e. ond this place here they were not removed.

Comment (zero-24) [reply language/en-GB/en-GB.com_users.ini]: This has just been patched by the last commit thanks! https://github.com/joomla/joomla-cms/pull/30787/commits/f36344a2a04693dcc56defd7fac6415c59f95e92

**Comment by richard67** [comment]
@zero-24 Could you update the branch of this PR to current staging? There was a problem with failing tests in staging when the branch of this PR was created, which meanwhile has been solved. That should make tests for this PR pass again, hopefully.

**Comment by zero-24** [comment]
> @zero-24 Could you update the branch of this PR to current staging? There was a problem with failing tests in staging when the branch of this PR was created, which meanwhile has been solved. That should make tests for this PR pass again, hopefully.

Done

**Comment by HLeithner** [comment]
@joomla/bug-squad can we have some tests here? I will create a RC tomorrow and I would like to have this merged.

**Comment by bayareajenn** [comment]
I have tested this item :red_circle: unsuccessfully on 01fea2ba8d34e9b103f9e8959b42259b776524db<br /><br />These items worked fine, test ok.
"install a site whith this patch applyed
create an non-superuser
try to reset the non-superuser
all should work correctly but with a different message"

These items didn't work for me.
"try to reset the superuser
there is now the same message but in the background a new mail is send to the super user telling him that he cant do what he wanted to do but how to do it.
enable joomla debug mode (and also make sure error reporting is active.
try to reset the super admin again
you now get the old error message but with the docs linked."

I never received an email telling me the steps I need to take to reset the super user pw. I don't think it's an email issue as the non-super user I set up DID send me verification codes via email.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30787">issues.joomla.org/tracker/joomla-cms/30787</a>.</sub>

**Comment by zero-24** [comment]
> I never received an email telling me the steps I need to take to reset the super user pw. I don't think it's an email issue as the non-super user I set up DID send me verification codes via email.

hmm but you did got the "old" message right?

**Comment by bayareajenn** [comment]
> 
> 
> > I never received an email telling me the steps I need to take to reset the super user pw. I don't think it's an email issue as the non-super user I set up DID send me verification codes via email.
> 
> hmm but you did got the "old" message right?

With the patch applied, I got the message: 

"Notice
Empty message body
If the email address you entered is registered on this site you will shortly receive an email with a link to reset the password for your account."

Never got the email.

**Comment by zero-24** [comment]
Hmm that does not make sense will have to debug that. Just in case it should be send to the mail that is set in the super user that you try to reset.

Can you see any errors in the php error log? Maybe i messed up something here.

**Comment by bayareajenn** [comment]
> Can you see any errors in the php error log? Maybe i messed up something here.

I don't have any idea how to check for errors in the php error log. I know that sucks. I did try and in error_log.txt there was nothing there. Maybe I'd need to talk to Rochen about it?

**Comment by zero-24** [comment]
Ok no Problem, will have to add to debug this to my todo list.

**Comment by zero-24** [comment]
please test again @bayareajenn took longer than expected for sure :D

**Comment by bayareajenn** [comment]
I have tested this item :white_check_mark: successfully on c4a13ef589239c61709376696c3802d799f4df7b<br /><br />It's working now. Got the new message with the patch applied:
"Notice

If the email address you entered is registered on this site you will shortly receive an email with a link to reset the password for your account."

Then I received the email stating:
"Hello,

A request has been made to reset your [SITE] account password. But, for security reasons, a Super User can't request a password reminder. Please contact another Super User or use an alternative method as described in the documentation: https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password0.000000"<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30787">issues.joomla.org/tracker/joomla-cms/30787</a>.</sub>

**Comment by brianteeman** [comment]
Is that the actual text of the email? If so then the link being generated is wrong

**Comment by bayareajenn** [comment]
> 
> 
> Is that the actual text of the email? If so then the link being generated is wrong

Yes, sir. That is the actual text of the email except that I removed the site title from it where it says [SITE].

**Comment by brianteeman** [comment]
Then that's a problem as thats not the url in the language string

`https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F`

**Comment by bayareajenn** [comment]
Shall I change my test result as unsuccessful then?

**Comment by bayareajenn** [comment]
Ok, so just in case, I re-fetched github and re-applied the patch, and re-tested and the email still says the same as reported above.

**Comment by zero-24** [comment]
hmm just looked into my test mail and I can confirm this behavior but I'm not sure what is happening here.

Maybe @infograf768 can help here to understand what broke JText to not use the URL from the language string? I already had to do this: https://github.com/joomla/joomla-cms/pull/30787/commits/c4a13ef589239c61709376696c3802d799f4df7b to get JText to use the full language string for some reason.

**Comment by brianteeman** [comment]
It could be the %3f at the end of the url (which is a ? character) or it could be something completely different.

Is there any reason for the url to be in the language string? Maybe you could try using sprintf and seeing if that solves it.

**Comment by zero-24** [comment]
> It could be the %3f at the end of the url (which is a ? character) or it could be something completely different.

Ah I haven't noticed that. Yup seems that was the issue as after moving it it works now. Thanks!

> Is there any reason for the url to be in the language string? Maybe you could try using sprintf and seeing if that solves it.

Not acutally but we usally have them in the strings but in this case it works the otherway too.

**Comment by brianteeman** [comment]
the url on the docs server does have the %3f

**Comment by zero-24** [comment]
> the url on the docs server does have the %3f

Yes and it has to as the page without the question mark does not exist: https://docs.joomla.org/How_do_you_recover_or_reset_your_admin_password/

But i have not noticed nor thougt that this could be a problem so i assumed something is wrong with JText :D

**Comment by brianteeman** [comment]
> Yes and it has to as the page without the question mark does not exist
That is what I was trying to say ;)

It's used without problem with JText in Joomla 4 with mod_login in the admin

**Comment by zero-24** [comment]
hmm maybe a combination with `/n/n` given that it worked partly when i used the `$interpretBackSlashes = false` switch.

File: components/com_users/models/reset.php [hunk 8]
```diff
@@ -424,6 +424,32 @@ public function processResetRequest($data)
...
+
+			$subject = JText::sprintf(
+				'COM_USERS_EMAIL_PASSWORD_RESET_SUBJECT',
+				$data['sitename']
+			);
+
+			$body = JText::sprintf(
+				'COM_USERS_REMIND_SUPERADMIN_BODY',
+				$data['sitename'],
+				'https://docs.joomla.org/Special:MyLanguage/How_do_you_recover_or_reset_your_admin_password%3F'
```
Comment (Quy) [inline components/com_users/models/reset.php:441]: The URL returns `404 Error - Page not found`.

**Comment by PhilETaylor** [comment]
It needs the `%` escaping with another `%`

The actual url NEEDS the %3F on the end of it https://docs.joomla.org/How_do_you_recover_or_reset_your_admin_password%3F

proof/test/example https://3v4l.org/bUlhg

**Comment by HLeithner** [comment]
I have the feeling there is not much interest in this pr...  I will create the RC today without a fix and tests it will not make it again in to 3,9

**Comment by CharlieH96** [comment]
This is really an issue because if someone gets your Email, there is plenty of ways to get the password for the Email address, then go and reset the users Joomla password.

For example; https://haveibeenpwned.com/

**Comment by brianteeman** [comment]
@CharlieH96 Please test this

**Comment by HLeithner** [comment]
This PR doesn't solve this for sites with public registration it's only a small security through obscurity enhancement .

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on 7d28ead2e0cc5a75d3495e68454ebfd6e530a4a6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30787">issues.joomla.org/tracker/joomla-cms/30787</a>.</sub>

**Comment by robinclapp** [comment]
I have tested this item :white_check_mark: successfully on 7d28ead2e0cc5a75d3495e68454ebfd6e530a4a6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30787">issues.joomla.org/tracker/joomla-cms/30787</a>.</sub>

**Comment by RickR2H** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30787">issues.joomla.org/tracker/joomla-cms/30787</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30787">issues.joomla.org/tracker/joomla-cms/30787</a>.</sub>

**Comment by laoneo** [comment]
Thanks!

**Pull request merged**